File: node/src/platforms.ts
import type { HttpClient } from './client.js';
import type { ListPlatformsResponse } from './types.js';

/**
 * Resource for inspecting platform availability.
 *
 * Access via `client.platforms`.
 *
 * Every social platform BulkPublish supports can be switched on or off server-side.
 * Use this resource to render an accurate "temporarily unavailable" or "connections
 * paused" state rather than assuming every platform can be connected or published to.
 *
 * Disabled platforms are always **included** in the response with `enabled: false`
 * and a `reason` — they are never omitted — so you can distinguish "switched off
 * right now" from "not supported".
 *
 * @example
 * ```typescript
 * const bp = new BulkPublish({ apiKey: 'bp_...' });
 *
 * const { platforms } = await bp.platforms.list();
 * for (const p of platforms) {
 *   if (!p.canPublish) console.log(`${p.displayName} is unavailable: ${p.message}`);
 * }
 * ```
 */
export class PlatformsResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * List every supported platform with its current availability.
   *
   * A platform in state `off` will reject post creation with a 403
   * `PLATFORM_DISABLED` error, and any already-scheduled posts targeting it are
   * **held** — not failed — until it is re-enabled. A platform in state
   * `connect_off` cannot accept new channel connections, but channels already
   * connected keep publishing normally.
   *
   * @returns List of platforms and their availability.
   *
   * @example
   * ```typescript
   * // Check before offering a connect button
   * const { platforms } = await bp.platforms.list();
   * const tumblr = platforms.find((p) => p.platform === 'tumblr');
   * if (tumblr?.canConnect) showConnectButton();
   * ```
   */
  list(): Promise<ListPlatformsResponse> {
    return this.http.get<ListPlatformsResponse>('/api/platforms');
  }
}

File: python/bulkpublish/platforms.py
"""
Platforms resource for the BulkPublish SDK.

Every social platform BulkPublish supports can be switched on or off
server-side. Read this before rendering connect buttons or assuming a post will
go out, so the UI can show "temporarily unavailable" instead of failing later.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List

if TYPE_CHECKING:
    from .client import _BaseClient

from .types import PlatformAvailability


class PlatformsResource:
    """Inspect platform availability.

    Access via ``client.platforms``:

    Example::

        bp = BulkPublish("bp_key")
        for p in bp.platforms.list()["platforms"]:
            if not p["canPublish"]:
                print(p["displayName"], "unavailable:", p.get("message"))
    """

    def __init__(self, client: _BaseClient) -> None:
        self._client = client

    def list(self) -> Dict[str, List[PlatformAvailability]]:
        """List every supported platform with its current availability.

        Disabled platforms are always **included** with ``enabled: False`` and a
        ``reason`` — never omitted — so you can tell "switched off right now"
        apart from "not supported".

        A platform in state ``off`` rejects post creation with a 403
        ``PLATFORM_DISABLED``, and posts already scheduled against it are
        **held**, not failed, until it is re-enabled. A platform in state
        ``connect_off`` accepts no new channel connections, but channels already
        connected keep publishing.

        A platform may also expose ``variants`` — sub-platforms gated separately
        because the vendor reviews them as a separate app. LinkedIn reports
        ``variants["organization"]`` for company pages, while the platform-level
        state describes personal profiles; check the variant before offering a
        company-page connect.

        Example::

            tumblr = next(
                p for p in bp.platforms.list()["platforms"] if p["platform"] == "tumblr"
            )
            if tumblr["canConnect"]:
                show_connect_button()
        """
        return self._client._request("GET", "/api/platforms")


class AsyncPlatformsResource:
    """Async version of :class:`PlatformsResource`.

    Example::

        async with AsyncBulkPublish("bp_key") as bp:
            platforms = await bp.platforms.list()
    """

    def __init__(self, client: _BaseClient) -> None:
        self._client = client

    async def list(self) -> Dict[str, List[PlatformAvailability]]:
        """List platform availability — see :meth:`PlatformsResource.list`."""
        return await self._client._request("GET", "/api/platforms")
